### PR TITLE
Add torch.jit.enabled

### LINKED ISSUE #12907: [jit] request: expose torch.jit._enabled

### PR DESCRIPTION
**Pull request by ssnl**
Fixes #12907

### CODE REVIEW

**Review by soumith — APPROVED** [review]

**Review by apaszke — CHANGES_REQUESTED** [review]
I don’t think that’s a good patch. It won’t disable the JIT entirely. This was only meant to be used with an env variable because it was not designed to change at runtime.

**Review by apaszke — CHANGES_REQUESTED** [review]
I don’t think that’s a good patch. It won’t disable the JIT entirely. This was only meant to be used with an env variable because it was not designed to change at runtime.

**Comment by apaszke** [comment]
I don’t think that’s a good patch. It won’t disable the JIT entirely. This was only meant to be used with an env variable because it was not designed to change at runtime.

I can’t request changes because GitHub is broken, but please don’t land this.

**Comment by apaszke** [comment]
I don’t think that’s a good patch. It won’t disable the JIT entirely. This was only meant to be used with an env variable because it was not designed to change at runtime.

I can’t request changes because GitHub is broken, but please don’t land this.

**Comment by apaszke** [comment]
I don’t think that’s a good patch. It won’t disable the JIT entirely. This was only meant to be used with an env variable because it was not designed to change at runtime.

I can’t request changes because GitHub is broken, but please don’t land this.

**Review by apaszke — CHANGES_REQUESTED** [review]
I don’t think that’s a good patch. It won’t disable the JIT entirely. This was only meant to be used with an env variable because it was not designed to change at runtime.

**Review by apaszke — CHANGES_REQUESTED** [review]
This was never meant to be changed at runtime, and will not do what people will expect.

**Review by apaszke — CHANGES_REQUESTED** [review]
This was never meant to be changed to runtime and won’t disable the JIT as the users will expect.

**Review by apaszke — CHANGES_REQUESTED** [review]
This wont fully disable the JIT and will be extremely misleading. This flag was never meant to be adjusted at run time.

**Comment by SsnL** [comment]
@apaszke I still have a question though. If it is unsafe, then why do we have `test_disabled` 
https://github.com/pytorch/pytorch/blob/df06fba1f12aa5415ce53e1d683df3fb5dc0e5d2/test/test_jit.py#L450-L470
which seems to rely on the assumption that it can be locally changed. Should we add a comment on the test?

**Comment by zdevito** [comment]
@SsnL The reason that test works is because it is carefully crafted to not hit the corner cases of disabling the JIT partially. I think the issues that @apaszke is talking about arise mostly when the JIT is enabled for a declaration/trace of something but disabled when that same thing is used (or vice versa).

**Comment by SsnL** [comment]
@zdevito I see. Thanks for the explanation! @apaszke also chatted offline with me about this, and helped me understand more about this code.

**Review by apaszke — COMMENTED** [review]
Can you please provide some rationale for this addition? I’m not sure why would anyone want to check that at run time really.

**Comment by SsnL** [comment]
@apaszke Sure.

1. Disabling JIT is helpful debugging. Being able to query if JIT is disabled or not further helps debugging.
2. Not everything is the same with JIT disabled/enabled. E.g., 

    1.  Not everything is supported in scripting.

    2. For traced functions, being traced when JIT is disabled or enabled causes semantics differences. E.g., `traced_net = trace(net.eval(), inp)` always gives the eval graph when JIT is enabled, but depends on `net`'s training mode otherwise.

    So I have different codepath depending on JIT is enabled or not. Therefore, I also want to be able to log whether JIT is enabled.

**Comment by apaszke** [comment]
> Not everything is supported in scripting.

Sure but would you really want to write a completely separate implementation of a function just because you might be disabling the JIT?

I really don’t think that anyone should have any logic in their programs that depend on the JIT being enabled or not. After all it’s a flag that’s supposed to make debugging easier, but if your program is taking different paths depending on this mode, then you’re not really inspecting the same thing.

**Comment by SsnL** [comment]
> would you really want to write a completely separate implementation of a function just because you might be disabling the JIT?

Why not give people the option to query the flag? It is so simple a change, and very reasonable too. Have you considered my points other than scripting?

And yes, in my first experience with JIT, I have two places that breaks when `PYTORCH_JIT=0`. One is because of the tracing behavior I described, and another is due to infinite recursion since the same function is returned at tracing. E.g., 
```py
def traced(x):
  return obj.meth(x)

obj.meth = types.MethodType(torch.jit.trace(traced, x), obj)
```

**Comment by apaszke** [comment]
Well your points are valid but instead of justifying adding this flag, they simply highlight bugs in the flag implementation that we should fix.

**Comment by SsnL** [comment]
closing in favor of https://github.com/pytorch/pytorch/issues/16293
